src/solvers/smt2/smt2_ast.h

```diff
@@ -0,0 +1,317 @@
+{
+public:
+  /// Empty constructor
+  smt2_astt() : tree_implementationt(irep_idt())
+  {
+  }
+
+  const irep_idt &id() const
+  {
+    return read().data;
+  }
+
+  explicit smt2_astt(irep_idt id) : tree_implementationt(std::move(id))
+  {
+  }
+};
+
+std::ostream &operator<<(std::ostream &stream, const smt2_astt &ast);
+
+class smt2_constantt : public smt2_astt
+{
+public:
+  /// Make an AST representing a integer literal
+  static smt2_constantt make(const mp_integer &m)
+  {
+    return smt2_constantt(integer2string(m));
+  }
+
+protected:
+  explicit smt2_constantt(irep_idt value) : smt2_astt(ID_constant)
+  {
+    write().named_sub = std::move(value);
+  }
+};
+
+class smt2_binary_literalt : public smt2_constantt
+{
+public:
+  /// Make an AST representing a integer literal
+  static smt2_binary_literalt make(const mp_integer &m)
+  {
+    return smt2_binary_literalt(integer2string(m, 2));
+  }
+
+  /// Takes a string in binary
+  explicit smt2_binary_literalt(const irep_idt &m)
+    : smt2_constantt("#b" + id2string(m))
+  {
+  }
+};
+
+class smt2_symbolt : public smt2_astt
+{
+public:
+  explicit smt2_symbolt(irep_idt symbol) : smt2_astt(ID_symbol)
+  {
+    write().named_sub = std::move(symbol);
+  }
+};
+
+class smt2_string_literalt : public smt2_astt
+{
+public:
+  explicit smt2_string_literalt(irep_idt literal)
+    : smt2_astt(ID_string_constant)
+  {
+    write().sub.emplace_back(std::move(literal));
+  }
+};
+
+class smt2_identifiert : public smt2_astt
+{
+public:
+  /// Identifiers are either symbols, or indexed: `( _ <symbol> <numeral>+ )`
+  explicit smt2_identifiert(smt2_symbolt sym) : smt2_astt(std::move(sym))
+  {
+  }
+
+  smt2_identifiert(smt2_symbolt sym, std::vector<smt2_constantt> indexes)
+    : smt2_astt(ID_identifier)
+  {
+    auto &subs = write().sub;
+    subs.emplace_back(std::move(sym));
+    std::move(indexes.begin(), indexes.end(), std::back_inserter(subs));
+  }
+};
+
+class smt2_sortt : public smt2_astt
+{
+public:
+  static smt2_sortt Bool()
+  {
+    return smt2_sortt(smt2_symbolt{"Bool"});
+  }
+
+  static smt2_sortt Int()
+  {
+    return smt2_sortt(smt2_symbolt{"Int"});
+  }
+
+  static smt2_sortt Real()
+  {
+    return smt2_sortt(smt2_symbolt{"Real"});
+  }
+
+  static smt2_sortt BitVec(const mp_integer &i)
+  {
+    return smt2_sortt{"BitVec", {smt2_constantt::make(i)}};
+  }
+
+  explicit smt2_sortt(smt2_symbolt symbol) : smt2_astt(std::move(symbol))
+  {
+  }
+
+  smt2_sortt(smt2_symbolt symbol, std::vector<smt2_astt> &&parameters)
+    : smt2_astt(ID_type)
+  {
+    auto &subs = write().sub;
+    subs.emplace_back(std::move(symbol));
+    std::move(parameters.begin(), parameters.end(), std::back_inserter(subs));
+  }
+
+private:
+  /// Indexed sort, only used for BitVec at the moment
+  smt2_sortt(irep_idt symbol, std::vector<smt2_constantt> &&parameters)
+    : smt2_astt(smt2_identifiert{smt2_symbolt(symbol), std::move(parameters)})
```

Comment:
there shouldn't be a need for this constructor if the one above is made to take a `smt2_identifiert`